`WebSocketServer` missing from TypeScript mapping

## Feature Request

### Description of Problem:

We are using Perspective to host tables in our backend TypeScript/node.js project and make them available by a `WebSocketServer`. I would prefer to use `import perspective from '@finos/perspective'` instead of `const perspective = require("@finos/perspective")` to gain all the TypeScript conveniences - however, VS Code/TypeScript complans that `perspective.WebSocketWorker` doesn't exist and the servers don't appear to be in the `index.d.ts` file.

### Potential Solutions:

Add type mappings to all the functions meant to be exposed to backend node.js users (this improves documentation and discoverability, also).

Thanks Tim. Unfortunately, this still doesn't work for me (which I find surprising for a docker build).

First minor gripe being the yarn build script seems to have a problem with trying to build "all":

```
ferris@andy-laptop:~/dev/perspective (master)$ PSP_DOCKER=1 yarn build
yarn run v1.22.4
$ node scripts/build.js
No .perspectiverc, running setup
? Focus (J)avascript, (P)ython, (A)ll? (all)
Invalid project "undefined" selected, running setup
? Focus (J)avascript, (P)ython, (A)ll? (all)
Invalid project "undefined" selected, running setup
? Focus (J)avascript, (P)ython, (A)ll? Javascript
-- Creating emsdk docker image
CMake Warning at CMakeLists.txt:182 (message):
  

  Building WASM binding

  Skipping C++ binding

  Skipping Python binding

  Building C++ with strict warnings

  Building RELEASE

  Skipping Perspective Documentation

...
```

And yet this still fails. After building the dependencies (i.e. a second attempt of the same) I get a problem with psp.async.js:

```
? Focus (J)avascript, (P)ython, (A)ll? Javascript
-- Creating emsdk docker image
CMake Warning at CMakeLists.txt:182 (message):
  

  Building WASM binding

  Skipping C++ binding

  Skipping Python binding

  Building C++ with strict warnings

  Building RELEASE

  Skipping Perspective Documentation


CMake Warning at CMakeLists.txt:45 (message):
  Dependency found - not rebuilding -
  /src/cpp/perspective/obj/rapidjson-build
Call Stack (most recent call first):
  CMakeLists.txt:214 (psp_build_dep)


-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- No Doxygen found. Documentation won't be built
-- Could NOT find GTestSrc (missing: GTEST_SOURCE_DIR GTEST_INCLUDE_DIR) 
CMake Warning at CMakeLists.txt:45 (message):
  Dependency found - not rebuilding -
  /src/cpp/perspective/obj/date-build
Call Stack (most recent call first):
  CMakeLists.txt:395 (psp_build_dep)


CMake Warning at CMakeLists.txt:45 (message):
  Dependency found - not rebuilding -
  /src/cpp/perspective/obj/hopscotch-build
Call Stack (most recent call first):
  CMakeLists.txt:396 (psp_build_dep)


CMake Warning at CMakeLists.txt:45 (message):
  Dependency found - not rebuilding -
  /src/cpp/perspective/obj/ordered-map-build
Call Stack (most recent call first):
  CMakeLists.txt:397 (psp_build_dep)


Building minimal Apache Arrow
CMake Warning at CMakeLists.txt:45 (message):
  Dependency found - not rebuilding -
  /src/cpp/perspective/obj/double-conversion-build
Call Stack (most recent call first):
  CMakeLists.txt:404 (psp_build_dep)


CMake Warning at CMakeLists.txt:45 (message):
  Dependency found - not rebuilding -
  /src/cpp/perspective/obj/arrow-build
Call Stack (most recent call first):
  CMakeLists.txt:405 (psp_build_dep)


/usr/local/bin/flatc
/usr/local/bin/flatc
Found Flatbuffers in /usr/local/include
-- Configuring done
-- Generating done
-- Build files have been written to: /src/cpp/perspective/obj
make: ['make', '-j12']
[  1%] Built target arrow_fb_files
[ 48%] Built target arrow
[ 98%] Built target psp
[100%] Built target perspective.async
-- Building psp.async.js
-- Copying WASM file psp.async.wasm
-- Creating wrapped js runtime
Error: Command failed: yarn run buildPath /home/ferris/dev/perspective
/home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:93
				throw err;
				^

Error: Cannot find module '@finos/perspective-webpack-plugin'
Require stack:
- /home/ferris/dev/perspective/packages/perspective/src/config/common.config.js
- /home/ferris/dev/perspective/packages/perspective/src/config/perspective.config.js
- /home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js
- /home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js
- /home/ferris/dev/perspective/node_modules/webpack/bin/webpack.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)
    at Function.Module._load (internal/modules/cjs/loader.js:690:27)
    at Module.require (internal/modules/cjs/loader.js:852:19)
    at require (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:161:20)
    at Object.<anonymous> (/home/ferris/dev/perspective/packages/perspective/src/config/common.config.js:3:27)
    at Module._compile (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Function.Module._load (internal/modules/cjs/loader.js:727:14)
    at Module.require (internal/modules/cjs/loader.js:852:19)
    at require (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:161:20)
    at Object.<anonymous> (/home/ferris/dev/perspective/packages/perspective/src/config/perspective.config.js:2:16)
    at Module._compile (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
    at Module.load (internal/modules/cjs/loader.js:815:32)
    at Function.Module._load (internal/modules/cjs/loader.js:727:14)
    at Module.require (internal/modules/cjs/loader.js:852:19)
    at require (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:161:20)
    at WEBPACK_OPTIONS (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:114:13)
    at requireConfig (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:116:6)
    at /home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:123:17
    at Array.forEach (<anonymous>)
    at module.exports (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:121:15)
    at /home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:71:45
    at Object.parse (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/yargs/yargs.js:567:18)
    at /home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:49:8
    at Object.<anonymous> (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:366:3)
    at Module._compile (internal/modules/cjs/loader.js:959:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)
    at Module.load (internal/modules/cjs/loader.js:815:32) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/ferris/dev/perspective/packages/perspective/src/config/common.config.js',
    '/home/ferris/dev/perspective/packages/perspective/src/config/perspective.config.js',
    '/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js',
    '/home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js',
    '/home/ferris/dev/perspective/node_modules/webpack/bin/webpack.js'
  ]
}
error Command failed with exit code 1.
ERROR: "build:webpack:umd" exited with 1.
error Command failed with exit code 1.
ERROR: "build:webpack" exited with 1.
error Command failed with exit code 1.
$ mkdirp dist && mkdirp obj
$ npm-run-all build:babel:* build:webpack build:webpack:umd:inline
$ babel src/js --source-maps --out-dir dist/esm
Successfully compiled 20 files with Babel.
$ BABEL_MODULE=auto babel dist/obj --source-maps --out-dir dist/cjs
Successfully compiled 1 file with Babel.
$ BABEL_MODULE=auto babel dist/esm --source-maps --out-dir dist/cjs
Successfully compiled 20 files with Babel.
$ npm-run-all -p build:webpack:* 
$ webpack --color --config src/config/perspective.config.js
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

    at /home/ferris/dev/perspective/node_modules/lerna/node_modules/execa/index.js:236:11
    at processTicksAndRejections (internal/process/task_queues.js:93:5) {
  code: 1,
  killed: false,
  stdout: '$ mkdirp dist && mkdirp obj\n' +
    '$ npm-run-all build:babel:* build:webpack build:webpack:umd:inline\n' +
    '$ babel src/js --source-maps --out-dir dist/esm\n' +
    'Successfully compiled 20 files with Babel.\n' +
    '$ BABEL_MODULE=auto babel dist/obj --source-maps --out-dir dist/cjs\n' +
    'Successfully compiled 1 file with Babel.\n' +
    '$ BABEL_MODULE=auto babel dist/esm --source-maps --out-dir dist/cjs\n' +
    'Successfully compiled 20 files with Babel.\n' +
    '$ npm-run-all -p build:webpack:* \n' +
    '$ webpack --color --config src/config/perspective.config.js\n' +
    'info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.\n' +
    'info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.\n' +
    'info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.\n',
  stderr: '/home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:93\n' +
    '\t\t\t\tthrow err;\n' +
    '\t\t\t\t^\n' +
    '\n' +
    "Error: Cannot find module '@finos/perspective-webpack-plugin'\n" +
    'Require stack:\n' +
    '- /home/ferris/dev/perspective/packages/perspective/src/config/common.config.js\n' +
    '- /home/ferris/dev/perspective/packages/perspective/src/config/perspective.config.js\n' +
    '- /home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js\n' +
    '- /home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js\n' +
    '- /home/ferris/dev/perspective/node_modules/webpack/bin/webpack.js\n' +
    '    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)\n' +
    '    at Function.Module._load (internal/modules/cjs/loader.js:690:27)\n' +
    '    at Module.require (internal/modules/cjs/loader.js:852:19)\n' +
    '    at require (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:161:20)\n' +
    '    at Object.<anonymous> (/home/ferris/dev/perspective/packages/perspective/src/config/common.config.js:3:27)\n' +
    '    at Module._compile (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)\n' +
    '    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)\n' +
    '    at Module.load (internal/modules/cjs/loader.js:815:32)\n' +
    '    at Function.Module._load (internal/modules/cjs/loader.js:727:14)\n' +
    '    at Module.require (internal/modules/cjs/loader.js:852:19)\n' +
    '    at require (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:161:20)\n' +
    '    at Object.<anonymous> (/home/ferris/dev/perspective/packages/perspective/src/config/perspective.config.js:2:16)\n' +
    '    at Module._compile (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)\n' +
    '    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)\n' +
    '    at Module.load (internal/modules/cjs/loader.js:815:32)\n' +
    '    at Function.Module._load (internal/modules/cjs/loader.js:727:14)\n' +
    '    at Module.require (internal/modules/cjs/loader.js:852:19)\n' +
    '    at require (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/v8-compile-cache/v8-compile-cache.js:161:20)\n' +
    '    at WEBPACK_OPTIONS (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:114:13)\n' +
    '    at requireConfig (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:116:6)\n' +
    '    at /home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:123:17\n' +
    '    at Array.forEach (<anonymous>)\n' +
    '    at module.exports (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js:121:15)\n' +
    '    at /home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:71:45\n' +
    '    at Object.parse (/home/ferris/dev/perspective/node_modules/webpack-cli/node_modules/yargs/yargs.js:567:18)\n' +
    '    at /home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:49:8\n' +
    '    at Object.<anonymous> (/home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js:366:3)\n' +
    '    at Module._compile (internal/modules/cjs/loader.js:959:30)\n' +
    '    at Object.Module._extensions..js (internal/modules/cjs/loader.js:995:10)\n' +
    '    at Module.load (internal/modules/cjs/loader.js:815:32) {\n' +
    "  code: 'MODULE_NOT_FOUND',\n" +
    '  requireStack: [\n' +
    "    '/home/ferris/dev/perspective/packages/perspective/src/config/common.config.js',\n" +
    "    '/home/ferris/dev/perspective/packages/perspective/src/config/perspective.config.js',\n" +
    "    '/home/ferris/dev/perspective/node_modules/webpack-cli/bin/utils/convert-argv.js',\n" +
    "    '/home/ferris/dev/perspective/node_modules/webpack-cli/bin/cli.js',\n" +
    "    '/home/ferris/dev/perspective/node_modules/webpack/bin/webpack.js'\n" +
    '  ]\n' +
    '}\n' +
    'error Command failed with exit code 1.\n' +
    'ERROR: "build:webpack:umd" exited with 1.\n' +
    'error Command failed with exit code 1.\n' +
    'ERROR: "build:webpack" exited with 1.\n' +
    'error Command failed with exit code 1.\n',
  failed: true,
  signal: null,
  cmd: 'yarn run build',
  timedOut: false,
  exitCode: 1
}
Command failed: lerna run build --loglevel silent 
(node:9758) UnhandledPromiseRejectionWarning: Error: Command failed: node scripts/build.js
    at checkExecSyncError (child_process.js:621:11)
    at execSync (child_process.js:657:15)
    at execute (/home/ferris/dev/perspective/scripts/setup.js:11:24)
    at Proxy.write (/home/ferris/dev/perspective/scripts/setup.js:36:17)
    at choose_project (/home/ferris/dev/perspective/scripts/setup.js:193:12)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:9758) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:9758) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
? Use docker for build env? Yes
Invalid project "undefined" selected, running setup
? Focus (J)avascript, (P)ython, (A)ll? python
-- Getting image for image: '' and python: 'python3.7'
EACCES: permission denied, open '/home/ferris/dev/perspective/python/perspective/dist/obj/arrow-src/.git/modules/cpp/submodules/parquet-testing/objects/02/f20a64c8b4a809aafe270ce6a0a91606e57d1d'
(node:10318) UnhandledPromiseRejectionWarning: Error: Command failed: node scripts/build.js
    at checkExecSyncError (child_process.js:621:11)
    at execSync (child_process.js:657:15)
    at execute (/home/ferris/dev/perspective/scripts/setup.js:11:24)
    at Proxy.write (/home/ferris/dev/perspective/scripts/setup.js:36:17)
    at choose_project (/home/ferris/dev/perspective/scripts/setup.js:193:12)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:10318) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:10318) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
Done in 2506.68s.
```

Any help would be greatly appreciated!

--- a/examples/remote-express-typescript/assets/index.css
+++ b/examples/remote-express-typescript/assets/index.css
@@ -1,0 +1,31 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2017, the Perspective Authors.
+ *
+ * This file is part of the Perspective library, distributed under the terms of
+ * the Apache License 2.0.  The full license can be found in the LICENSE file.
+ *
+ */
+
+ perspective-viewer {
+    position: absolute;
+    top: 0;
+    left: 0;
+    right: 0;
+    bottom: 0;
+}
+
+@media (max-width: 600px) {
+    html {
+        overflow: hidden;
+    }
+
+    body {
+        position: fixed;
+        height: 100%;
+        width: 100%;
+        margin: 0;
+        overflow: hidden;
+        touch-action: none;
+    }
+}

--- a/examples/remote-express-typescript/assets/index.html
+++ b/examples/remote-express-typescript/assets/index.html
@@ -1,0 +1,69 @@
+<!--
+   
+   Copyright (c) 2017, the Perspective Authors.
+   
+   This file is part of the Perspective library, distributed under the terms of
+   the Apache License 2.0.  The full license can be found in the LICENSE file.
+
+-->
+
+<!DOCTYPE html>
+<html>
+
+<head>
+
+    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, minimum-scale=1, user-scalable=no">
+
+    <script src="perspective-viewer.js"></script>
+    <script src="perspective-viewer-hypergrid.js"></script>
+    <script src="perspective-viewer-d3fc.js"></script>
+
+    <script src="perspective.js"></script>
+
+    <link rel='stylesheet' href="index.css">
+    <link rel='stylesheet' href="material.dark.css">
+
+</head>
+
+<body>
+
+    <perspective-viewer 
+        id="view1"
+        plugin="d3_heatmap"
+        row-pivots='["client"]'
+        columns='["chg"]'
+        column-pivots='["name"]'>
+    
+    </perspective-viewer>
+
+    <script>
+
+        window.addEventListener('WebComponentsReady', async function() {
+
+            // Create two perspective interfaces, one remotely via WebSocket, 
+            // and one local via WebWorker.
+            
+            const websocket = perspective.websocket(`${window.location.origin.replace("http", "ws")}/subscribe`);
+            const worker = perspective.worker();
+            worker.initialize_profile_thread();
+
+            // Get a proxy for a view named "data_source_one", registered on
+            // the server with a reciprocal call to `host_view()`.
+            // No data is transferred, `view` is a virtual handle for data on
+            // the server.
+            const table = websocket.open_table('remote_table');
+
+            // Create a `table` from this, owned by the local WebWorker.
+            // Data is transferred from `view` to the local WebWorker, both
+            // the current state and all future updates, as Arrows.
+            // const table = worker.table(view, {limit: 10000});
+            
+            // Load this in the `<perspective-viewer>`.
+            document.getElementById('view1').load(table);
+        });
+
+    </script>
+
+</body>
+
+</html>

--- a/examples/remote-express-typescript/package.json
+++ b/examples/remote-express-typescript/package.json
@@ -1,0 +1,21 @@
+{
+    "name": "remote-express-typescript",
+    "private": true,
+    "version": "0.4.5",
+    "description": "An example of 2 Perspectives, one client and one server, streaming via Apache Arrow.",
+    "scripts": {
+        "start": "tsc && node dist/server.js"
+    },
+    "keywords": [],
+    "license": "Apache-2.0",
+    "dependencies": {
+        "@finos/perspective": "^0.4.5",
+        "@finos/perspective-viewer": "^0.4.5",
+        "@finos/perspective-viewer-d3fc": "^0.4.5",
+        "@finos/perspective-viewer-hypergrid": "^0.4.5"
+    },
+    "devDependencies": {
+        "@types/express": "^4.17.3",
+        "@types/express-ws": "^3.0.0"
+    }
+}

--- a/examples/remote-express-typescript/src/server.ts
+++ b/examples/remote-express-typescript/src/server.ts
@@ -1,0 +1,28 @@
+/******************************************************************************
+ *
+ * Copyright (c) 2017, the Perspective Authors.
+ *
+ * This file is part of the Perspective library, distributed under the terms of
+ * the Apache License 2.0.  The full license can be found in the LICENSE file.
+ *
+ */
+
+import {WebSocketManager, perspective_assets, Table} from "@finos/perspective";
+import path from "path";
+import express from "express";
+import expressWs from "express-ws";
+import {AddressInfo} from "net";
+
+// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
+// @ts-ignore
+import {securities} from "../../datasources";
+
+const app = expressWs(express()).app;
+
+const manager = new WebSocketManager();
+securities().then((table: Table) => manager.host_table("remote_table", table));
+
+app.ws("/subscribe", ws => manager.add_connection(ws));
+app.use("/", perspective_assets([path.resolve(__dirname, "../assets")], true));
+
+const server = app.listen(8080, () => console.log(`Listening on port ${(server.address() as AddressInfo).port}`));

--- a/examples/remote-express-typescript/tsconfig.json
+++ b/examples/remote-express-typescript/tsconfig.json
@@ -1,0 +1,21 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "esModuleInterop": true,
+        "allowSyntheticDefaultImports": true,
+        "target": "es6",
+        "noImplicitAny": true,
+        "moduleResolution": "node",
+        "sourceMap": true,
+        "outDir": "./dist",
+        "baseUrl": ".",
+        "strict": true,
+        "alwaysStrict": true,
+    },
+    "include": [
+        "src/server.ts"
+    ],
+    "exclude": [
+        "node_modules",
+    ]
+}

--- a/package.json
+++ b/package.json
@@ -29,6 +29,7 @@
         "@babel/plugin-transform-runtime": "^7.8.3",
         "@babel/polyfill": "^7.8.3",
         "@babel/preset-env": "^7.8.4",
+        "@types/ws": "^7.2.2",
         "@typescript-eslint/eslint-plugin": "^2.4.0",
         "@typescript-eslint/parser": "^2.4.0",
         "arraybuffer-loader": "^1.0.2",

--- a/packages/perspective/index.d.ts
+++ b/packages/perspective/index.d.ts
@@ -1,4 +1,6 @@
 declare module "@finos/perspective" {
+    import * as ws from "ws";
+
     /**** object types ****/
     export enum TypeNames {
         STRING = "string",
@@ -123,7 +125,7 @@ declare module "@finos/perspective" {
         schema(): Promise<Schema>;
         size(): Promise<number>;
         update(data: TableData): void;
-        view(config: ViewConfig): View;
+        view(config?: ViewConfig): View;
     };
 
     /**** perspective ****/
@@ -135,19 +137,44 @@ declare module "@finos/perspective" {
         table(data: TableData, options?: TableOptions): Table;
     };
 
-    export type Client = {
+    export class WebSocketClient {
         open_table(name: string): Table;
         open_view(name: string): View;
+        terminate(): void;
+        initialize_profile_thread(): void;
+        send(msg: any): void;
+    }
+
+    export type WebSocketServerOptions = {
+        assets?: string[];
+        host_psp?: boolean;
+        port?: number;
+        on_start?: any;
     };
+
+    export class WebSocketManager {
+        add_connection(ws: ws): void;
+        host_table(name: string, table: Table): void;
+        host_view(name: string, view: View): void;
+        eject_table(name: string): void;
+        eject_view(name: string): void;
+    }
+
+    export class WebSocketServer extends WebSocketManager {
+        constructor(config?: WebSocketServerOptions);
+        close(): void;
+    }
+
+    export function perspective_assets(assets: string[], host_psp: boolean): (request: any, response: any) => void;
 
     type perspective = {
         TYPE_AGGREGATES: ValuesByType;
         TYPE_FILTERS: ValuesByType;
         SORT_ORDERS: SortOrders;
-        table(data_or_schema: TableData | Schema, options: TableOptions): Table;
+        table(data_or_schema: TableData | Schema, options?: TableOptions): Table;
         worker(): PerspectiveWorker;
         shared_worker(): PerspectiveWorker;
-        websocket(url: string): Client;
+        websocket(url: string): WebSocketClient;
         override: (x: any) => void;
     };
 

--- a/packages/perspective/index.d.ts
+++ b/packages/perspective/index.d.ts
@@ -1,85 +1,85 @@
-declare module '@finos/perspective' {
+declare module "@finos/perspective" {
     /**** object types ****/
     export enum TypeNames {
-        STRING = 'string',
-        FLOAT = 'float',
-        INTEGER = 'integer',
-        BOOLEAN = 'boolean',
-        DATE = 'date'
+        STRING = "string",
+        FLOAT = "float",
+        INTEGER = "integer",
+        BOOLEAN = "boolean",
+        DATE = "date"
     }
 
     export type ValuesByType = {
-        [ key in TypeNames ]: Array<string>
-    }
+        [key in TypeNames]: Array<string>;
+    };
 
     export type ValueByType = {
-        TypeNames: Array<string>
-    }
+        TypeNames: Array<string>;
+    };
 
     export enum SortOrders {
-        ASC = 'asc',
-        ASC_ABS = 'asc abs',
-        DESC = 'desc',
-        DESC_ABS = 'desc abs',
-        NONE = 'none',
+        ASC = "asc",
+        ASC_ABS = "asc abs",
+        DESC = "desc",
+        DESC_ABS = "desc abs",
+        NONE = "none"
     }
 
     enum NUMBER_AGGREGATES {
-        ANY = 'any',
-        AVERAGE = 'avg',
-        COUNT = 'count',
-        DISTINCT_COUNT = 'distinct count',
-        DOMINANT = 'dominant',
-        FIRST = 'first',
-        LAST = 'last',
-        HIGH = 'high',
-        LOW = 'low',
-        MEAN = 'mean',
-        MEDIAN = 'median',
-        PCT_SUM_PARENT = 'pct sum parent',
-        PCT_SUM_TOTAL = 'pct sum grand total',
-        SUM = 'sum',
-        SUM_ABS = 'sum abs',
-        SUM_NOT_NULL = 'sum not null',
-        UNIQUE = 'unique'
+        ANY = "any",
+        AVERAGE = "avg",
+        COUNT = "count",
+        DISTINCT_COUNT = "distinct count",
+        DOMINANT = "dominant",
+        FIRST = "first",
+        LAST = "last",
+        HIGH = "high",
+        LOW = "low",
+        MEAN = "mean",
+        MEDIAN = "median",
+        PCT_SUM_PARENT = "pct sum parent",
+        PCT_SUM_TOTAL = "pct sum grand total",
+        SUM = "sum",
+        SUM_ABS = "sum abs",
+        SUM_NOT_NULL = "sum not null",
+        UNIQUE = "unique"
     }
 
     enum STRING_AGGREGATES {
-        ANY = 'any',
-        COUNT = 'count',
-        DISTINCT_COUNT = 'distinct count',
-        DISTINCT_LEAF = 'distinct leaf',
-        DOMINANT = 'dominant',
-        FIRST = 'first',
-        LAST = 'last',
-        UNIQUE = 'unique'
+        ANY = "any",
+        COUNT = "count",
+        DISTINCT_COUNT = "distinct count",
+        DISTINCT_LEAF = "distinct leaf",
+        DOMINANT = "dominant",
+        FIRST = "first",
+        LAST = "last",
+        UNIQUE = "unique"
     }
 
     enum BOOLEAN_AGGREGATES {
-        AND = 'and',
-        ANY = 'any',
-        COUNT = 'count',
-        DISTINCT_COUNT = 'distinct count',
-        DISTINCT_LEAF = 'distinct leaf',
-        DOMINANT = 'dominant',
-        FIRST = 'first',
-        LAST = 'last',
-        OR ='or',
-        UNIQUE = 'unique'
+        AND = "and",
+        ANY = "any",
+        COUNT = "count",
+        DISTINCT_COUNT = "distinct count",
+        DISTINCT_LEAF = "distinct leaf",
+        DOMINANT = "dominant",
+        FIRST = "first",
+        LAST = "last",
+        OR = "or",
+        UNIQUE = "unique"
     }
 
     /**** Schema ****/
     type SchemaType = TypeNames | NUMBER_AGGREGATES | STRING_AGGREGATES | BOOLEAN_AGGREGATES;
 
     export type Schema = {
-        [ key: string ]: TypeNames ;
-    }
+        [key: string]: TypeNames;
+    };
 
     export interface SerializeConfig {
-        start_row: number,
-        end_row: number,
-        start_col: number,
-        end_col: number,
+        start_row: number;
+        end_row: number;
+        start_col: number;
+        end_col: number;
     }
 
     /**** View ****/
@@ -90,27 +90,27 @@ declare module '@finos/perspective' {
         on_delete(callback: Function): void;
         on_update(callback: UpdateCallback): void;
         schema(): Promise<Schema>;
-        to_arrow(options?: SerializeConfig & { data_slice: any }): Promise<ArrayBuffer>;
+        to_arrow(options?: SerializeConfig & {data_slice: any}): Promise<ArrayBuffer>;
         to_columns(options?: SerializeConfig): Promise<Array<object>>;
-        to_csv(options?: SerializeConfig & { config: object }): Promise<string>;
+        to_csv(options?: SerializeConfig & {config: object}): Promise<string>;
         to_json(options?: SerializeConfig): Promise<Array<object>>;
-    }
+    };
 
     /**** Table ****/
-    export type UpdateCallback = (data: Array<object>) => void
+    export type UpdateCallback = (data: Array<object>) => void;
 
-    export type TableData = string | Array<object> | { [key: string]: Array<object> } | { [key: string]: string }
+    export type TableData = string | Array<object> | {[key: string]: Array<object>} | {[key: string]: string};
 
     export type TableOptions = {
-        index?: string,
-        limit?: number
-    }
+        index?: string;
+        limit?: number;
+    };
 
     export type ViewConfig = {
         columns?: Array<string>;
         row_pivots?: Array<string>;
         column_pivots?: Array<string>;
-        aggregates?: { [column_name:string]: string; },
+        aggregates?: {[column_name: string]: string};
         sort?: Array<Array<string>>;
         filter?: Array<Array<string>>;
     };
@@ -124,33 +124,32 @@ declare module '@finos/perspective' {
         size(): Promise<number>;
         update(data: TableData): void;
         view(config: ViewConfig): View;
-    }
+    };
 
     /**** perspective ****/
     export enum PerspectiveEvents {
-        PERSPECTIVE_READY = 'perspective-ready'
+        PERSPECTIVE_READY = "perspective-ready"
     }
 
     export type PerspectiveWorker = {
         table(data: TableData, options?: TableOptions): Table;
-    }
-
+    };
 
     export type Client = {
-        open_table(name: string): Table,
-        open_view(name: string): View,
-    }
+        open_table(name: string): Table;
+        open_view(name: string): View;
+    };
 
     type perspective = {
-        TYPE_AGGREGATES: ValuesByType,
-        TYPE_FILTERS: ValuesByType,
-        SORT_ORDERS: SortOrders,
-        table(data_or_schema : TableData | Schema, options: TableOptions): Table,
-        worker(): PerspectiveWorker,
-        shared_worker(): PerspectiveWorker,
+        TYPE_AGGREGATES: ValuesByType;
+        TYPE_FILTERS: ValuesByType;
+        SORT_ORDERS: SortOrders;
+        table(data_or_schema: TableData | Schema, options: TableOptions): Table;
+        worker(): PerspectiveWorker;
+        shared_worker(): PerspectiveWorker;
         websocket(url: string): Client;
-        override: (x: any) => void
-    }
+        override: (x: any) => void;
+    };
 
     const impl: perspective;
 
@@ -169,4 +168,3 @@ declare module "@finos/perspective/build/psp.sync.wasm" {
 
 declare module "@finos/perspective/build/perspective.wasm.worker.js" {}
 declare module "@finos/perspective/build/perspective.asmjs.worker.js" {}
-

